## Mini-Project-1 
**Name: Darshit Jagetiya, Student ID: 014517932**

**->**  From the analysis performed below, it is found that the engagement is decreased recently in the last month due to some of the reasons. 
To find the reason, various operations and queries have been used considering different columns and their relation with the engagement.

**->**  The operations has been performed in the following manner considering each of the factor for the decrease in engagement. 

**->Range of dataset: **  Found the first and last date (range) for each of the tables whose data is given to perform the analysis efficiently.

**->Is engagement decreasing?:**  Found if the number of engagements are decreasing recently. Also, checked if only the number of enagagements is decreasing or even number of events for signup_flow has been reduced as well.

**-> Reason 1:**  Analyzed the email data set to find the reason for the decrement in the the engagements and not signup_flow.

**-> Reason 2:**  Checked for (top 10) companies who has maximum engagements in the last two months.

**-> Reason 3:**  There is a possibility that eLinks have a competitor company which is offering the same services. We can check the services offered by the eLinks and check if there is any particular feature which is not used by the users or all the features are not being used.

Link for databricks project for visualizing along with graphs :
https://databricks-prod-cloudfront.cloud.databricks.com/public/4027ec902e239c93eaaa8714f173bcfc/3007571033566758/3282462353226895/1538677595945462/latest.html

In [ ]:
# File location and type
file_location = "/FileStore/tables/users.csv"
file_type = "csv"

# CSV options
infer_schema = "false"
first_row_is_header = "true"
delimiter = ","

# The applied options are for CSV files. For other file types, these will be ignored.
df = spark.read.format(file_type) \
  .option("inferSchema", infer_schema) \
  .option("header", first_row_is_header) \
  .option("sep", delimiter) \
  .load(file_location)

display(df)

In [ ]:

# File location and type
file_location2 = "/FileStore/tables/emails.csv"
file_type = "csv"

# CSV options
infer_schema = "false"
first_row_is_header = "true"
delimiter = ","

# The applied options are for CSV files. For other file types, these will be ignored.
df2 = spark.read.format(file_type) \
  .option("inferSchema", infer_schema) \
  .option("header", first_row_is_header) \
  .option("sep", delimiter) \
  .load(file_location2)

display(df2)

In [ ]:
# File location and type
file_location3 = "/FileStore/tables/events.csv"
file_type = "csv"

# CSV options
infer_schema = "false"
first_row_is_header = "true"
delimiter = ","

# The applied options are for CSV files. For other file types, these will be ignored.
df3 = spark.read.format(file_type) \
  .option("inferSchema", infer_schema) \
  .option("header", first_row_is_header) \
  .option("sep", delimiter) \
  .load(file_location3)

display(df3)

In [5]:
# Create a view or table

temp_table_name = "users_csv"

df.createOrReplaceTempView(temp_table_name)

temp_table_name2="emails_csv"
df2.createOrReplaceTempView(temp_table_name2)

temp_table_name3="events_csv"
df3.createOrReplaceTempView(temp_table_name3)

# Analyzing the datasets

#### Range of dataset:
Found the first and last date (range) for each of the tables whose data is given to perform the analysis efficiently.

In [9]:
%sql
SELECT min(created_at),max(created_at) FROM users_csv;

min(created_at) max(created_at) 2013-01-01 04:38:30 2014-08-31 20:40:32

In [10]:
%sql
SELECT min(occurred_at),max(occurred_at) FROM events_csv;

min(occurred_at) max(occurred_at) 2014-05-01 00:54:09 2014-08-31 20:40:32

In [11]:
%sql
SELECT min(occurred_at),max(occurred_at) FROM emails_csv;

min(occurred_at) max(occurred_at) 2014-05-01 09:28:25 2014-08-31 23:00:42

#### Number of total, events, engagements events and signu_flow events.
Found if the number of engagements are decreasing recently. Also, checked if only the number of enagagements is decreasing or even number of events for signup_flow has been reduced as well.

In [13]:
%sql
SELECT DATE_TRUNC('month',ev.occurred_at) as month, COUNT(*) AS total_events,
COUNT(CASE WHEN event_type="engagement" THEN u.user_id END) AS engaged_event
FROM users_csv u join events_csv ev using (user_id)
where ev.occurred_at>'2014-05-01' 
group by 1
order by 1;
--The total events is decreased and the engaging events is also decreasing.

month total_events engaged_event 2014-05-01T00:00:00.000+0000 39234 35078 2014-06-01T00:00:00.000+0000 37357 32764 2014-07-01T00:00:00.000+0000 41267 36055 2014-08-01T00:00:00.000+0000 32142 26846

In [14]:
%sql
SELECT DATE_TRUNC('month',ev.occurred_at) as month, count(*) AS signup_flow_events
FROM  events_csv ev 
where event_type="signup_flow" and (ev.occurred_at>'2014-04' and ev.occurred_at<'2014-09')
group by 1
order by 1;
--This shows that though, the number of engaging and total_event is decreasing, signup_flow event_type is increasing in the last month.

month signup_flow_events 2014-05-01T00:00:00.000+0000 4156 2014-06-01T00:00:00.000+0000 4593 2014-07-01T00:00:00.000+0000 5212 2014-08-01T00:00:00.000+0000 5296

In [ ]:
%sql
SELECT DATE_TRUNC('day',ev.occurred_at) as day, COUNT(*) AS total_events,
COUNT(CASE WHEN event_type="engagement" THEN u.user_id END) AS engaged_events
FROM users_csv u join events_csv ev using (user_id)
where ev.occurred_at>'2014-05-01' 
group by 1
order by 1;

In [ ]:
%sql
SELECT DATE_TRUNC('day',ev.occurred_at) as day, count(*) AS signup_flow_events
FROM  events_csv ev 
where event_type="signup_flow" and (ev.occurred_at>'2014-04' and ev.occurred_at<'2014-09')
group by 1
order by 1;

#### Thus, from the above bar graphs and tables it is clear that the number of engagements is decreasing recently.

### Analysing email dataset (Reason 1)
Here, I have found one of the reason which has impacted the decrement in user engagement.

In [19]:
%sql
SELECT DATE_TRUNC('month',occurred_at) as month,count(*) as emails
FROM emails_csv 
group by month
order by month;
--sent_weekly_digestemail_openemail_clickthrough


month emails 2014-05-01T00:00:00.000+0000 18723 2014-06-01T00:00:00.000+0000 20976 2014-07-01T00:00:00.000+0000 25167 2014-08-01T00:00:00.000+0000 25523

In [20]:
%sql
SELECT DATE_TRUNC('month',occurred_at) as month,count(*) as weekly_digest
FROM emails_csv 
where action="sent_weekly_digest"
GROUP BY month
ORDER BY month;
--sent_weekly_digestemail_openemail_clickthrough


month weekly_digest 2014-05-01T00:00:00.000+0000 11730 2014-06-01T00:00:00.000+0000 13155 2014-07-01T00:00:00.000+0000 15902 2014-08-01T00:00:00.000+0000 16480

In [21]:
%sql
SELECT DATE_TRUNC('month',occurred_at) as month,count(*) as emails_open
FROM emails_csv 
where action="email_open"
GROUP BY month
ORDER BY month;

month emails_open 2014-05-01T00:00:00.000+0000 4212 2014-06-01T00:00:00.000+0000 4658 2014-07-01T00:00:00.000+0000 5611 2014-08-01T00:00:00.000+0000 5978

In [22]:
%sql
SELECT DATE_TRUNC('month',occurred_at) as month,count(*) as emails_clickthrough
FROM emails_csv 
where action="email_clickthrough"
GROUP BY month
ORDER BY month;

month emails_clickthrough 2014-05-01T00:00:00.000+0000 2023 2014-06-01T00:00:00.000+0000 2274 2014-07-01T00:00:00.000+0000 2721 2014-08-01T00:00:00.000+0000 1992

### Reason 1 for decrease in engagements:
The first and the foremost reason is due to the weekly_digest sent through the emails. The weekly_digest sent in the month 5 and 6 are very low comparatively to the 7 and 8 month. As the number of weekly digest are increasing in the last 2 months, the number of actions that users have opened the emails is increased but number of clickthroughs are not increased. This shows that too frequent weekly_digest sent to the users have decreased the clickthrough as the weekly_digest sent to users is very trivial. In month 5 and 6, with less weekly_digest sent to the user, still the number of email_clickthrough is found to be higher than the last month.



Weekly digest: 2014-05 -> 18723,  2014-06 -> 20976,   2014-07 -> 25167,   2014-08 -> 25523

Email open: 2014-05 ->11730,  2014-06 ->13155,  2014-07 ->15902,  2014-08 ->16480

Email clickthrough: 2014-05 ->2023, 2014-06 ->2274, 2014-07 ->2721, 2014-08 ->1992

The above reason can be more clearly explained if data in emails table was given for the first 4 months of 2014. If I was working for the company, then I would ask for more data for the above table from my company.


### Analyzing for companies (Reason 2):
  First we will check if the number of companies using eLinks have decreased. Then we will check if the number of engagements per company have decreased or not.

In [25]:
%sql
SELECT count(DISTINCT(company_id)) as total_number_of_companies from users_csv;

total_number_of_companies 13198

In [26]:
%sql
SELECT DATE_TRUNC('month',ev.occurred_at) as month,count(DISTINCT(company_id)) as number_of_companies
FROM users_csv u join events_csv ev using (user_id) 
where ev.event_type="engagement" and ev.occurred_at>='2014-05-01'
group by 1
order by 1;

month number_of_companies 2014-05-01T00:00:00.000+0000 1169 2014-06-01T00:00:00.000+0000 1270 2014-07-01T00:00:00.000+0000 1364 2014-08-01T00:00:00.000+0000 1269

#### The above table shows that the number of companies using eLinks have not been decreasing per month. But only their engagement can be decreasing which is checked in the next step. We can check it for only the first 15 companies which will make a significant impact.

In [28]:
%sql
SELECT u.company_id, count(*) as number_of_engagements
FROM users_csv u JOIN events_csv ev USING (user_id)
WHERE event_type="engagement" and (ev.occurred_at>'2014-06-30' and ev.occurred_at<'2014-08-01')
GROUP BY 1
ORDER BY 2 desc Limit 15;

company_id number_of_engagements 1.0 2156 2.0 1283 6.0 792 3.0 728 4.0 497 13.0 433 7.0 388 10.0 363 8.0 344 5.0 332 9.0 262 6789.0 253 24.0 235 44.0 197 255.0 194

In [29]:
%sql
SELECT u.company_id, count(*) number_of_engagements
FROM users_csv u JOIN events_csv ev USING (user_id)
WHERE event_type="engagement" and (ev.occurred_at>'2014-07-31' and ev.occurred_at<'2014-09-01')
GROUP BY 1
ORDER BY 2 desc Limit 15;

company_id number_of_engagements 1.0 1256 3.0 1113 2.0 484 4.0 429 8.0 384 13.0 286 6789.0 269 5343.0 264 10.0 196 5.0 194 7.0 188 1242.0 187 123.0 186 19.0 185 106.0 181

### Reason 2 for decrease in engagements:
From the above analysis, we found that the number of companies using eLinks have not decreased significantly but only the number of engagements per company has decreased.
This can be happening due to the decrease in  the workload of the employees of the company. We can check for the 15 companies which has highest user enagagement for the last 2 months. 

It can be seen that for company_id 1, which had 2156 user enagagements in July (2014-07), has only 1256 engagements in August (2014-07). Similarly, for company_id 2, the engagements has dropped from 1283 to 484 from July to August.

#### To check if the number of users has decreased in the last month which can be the reason for decrease in engagement rather than workload of the users.

In [32]:
%sql
SELECT count(DISTINCT(u.user_id)) as total_user_id
from users_csv u;

total_user_id 19066

In [33]:
%sql
SELECT count(DISTINCT(ev.user_id)) as userid_appearing_in_events
FROM  events_csv ev 
where ev.event_type="engagement" and ev.occurred_at>'2014-04-01' ;

userid_appearing_in_events 4878

In [34]:
%sql
SELECT count(DISTINCT(ev.user_id)) as distinct_users,DATE_TRUNC('month',ev.occurred_at) as month
FROM  events_csv ev 
where ev.event_type="engagement" and ev.occurred_at>='2014-05-01'
group by 2
order by 2;

--This shows that the number of users are not decreased significanlty in the last month, but only their engagement is decreased.
--Thus, decrement in number of users is not the reason for decrement in engagement.

distinct_users month 1494 2014-05-01T00:00:00.000+0000 1614 2014-06-01T00:00:00.000+0000 1764 2014-07-01T00:00:00.000+0000 1611 2014-08-01T00:00:00.000+0000

#### This proves that our reason 2 for decrease in workload is still valid as number of users are not decreasing. Our output shows that the number of users have not decreased significantly but only engagement per user has been decreased.

### Is the userinterface getting bad for eLinks and have a competitor providing same services. (Reason 3)
Check if all the engagement events are decreasing or only few of them are decreasing. Based on that we can decide if user interface of eLinks has got bad because though more number of people are signing_up, they are not using their eLinks after that.

In [37]:
%sql
SELECT  DATE_TRUNC('month',occurred_at) as month,count(event_name) as home_page_engagements
FROM events_csv 
WHERE event_name in ('home_page')
GROUP BY month
ORDER BY month;

month home_page_engagements 2014-05-01T00:00:00.000+0000 10319 2014-06-01T00:00:00.000+0000 9684 2014-07-01T00:00:00.000+0000 10589 2014-08-01T00:00:00.000+0000 7922

In [38]:
%sql
SELECT  DATE_TRUNC('month',occurred_at) as month,count(event_name) as like_message_engagements
FROM events_csv
WHERE event_name in ('like_message')
GROUP BY month
ORDER BY month;

month like_message_engagements 2014-05-01T00:00:00.000+0000 6723 2014-06-01T00:00:00.000+0000 6118 2014-07-01T00:00:00.000+0000 6661 2014-08-01T00:00:00.000+0000 4913

In [39]:
%sql
SELECT  DATE_TRUNC('month',occurred_at) as month,count(event_name) as login_engagements
FROM events_csv 
WHERE event_name in ('login')
GROUP BY month
ORDER BY month;

month login_engagements 2014-05-01T00:00:00.000+0000 3939 2014-06-01T00:00:00.000+0000 3835 2014-07-01T00:00:00.000+0000 4185 2014-08-01T00:00:00.000+0000 3293

In [40]:
%sql
SELECT  DATE_TRUNC('month',occurred_at) as month,count(event_name) as search_autocomplete_engagements
FROM events_csv 
WHERE event_name in ('search_autocomplete')
GROUP BY month
ORDER BY month;

month search_autocomplete_engagements 2014-05-01T00:00:00.000+0000 1692 2014-06-01T00:00:00.000+0000 1732 2014-07-01T00:00:00.000+0000 1888 2014-08-01T00:00:00.000+0000 1619

In [41]:
%sql
SELECT  DATE_TRUNC('month',occurred_at) as month,count(event_name) as search_run_engagements
FROM events_csv 
WHERE event_name in ('search_run')
GROUP BY month
ORDER BY month;

month search_run_engagements 2014-05-01T00:00:00.000+0000 1458 2014-06-01T00:00:00.000+0000 1265 2014-07-01T00:00:00.000+0000 1500 2014-08-01T00:00:00.000+0000 965

In [42]:
%sql
SELECT  DATE_TRUNC('month',occurred_at) as month,count(event_name) as send_message_engagements
FROM events_csv 
WHERE event_name in ('send_message')
GROUP BY month
ORDER BY month;

month send_message_engagements 2014-05-01T00:00:00.000+0000 3731 2014-06-01T00:00:00.000+0000 3516 2014-07-01T00:00:00.000+0000 3716 2014-08-01T00:00:00.000+0000 2640

In [43]:
%sql
SELECT  DATE_TRUNC('month',occurred_at) as month,count(event_name) as view_inbox_engagements
FROM events_csv 
WHERE event_name in ('view_inbox')
GROUP BY month
ORDER BY month;

month view_inbox_engagements 2014-05-01T00:00:00.000+0000 6003 2014-06-01T00:00:00.000+0000 5684 2014-07-01T00:00:00.000+0000 6376 2014-08-01T00:00:00.000+0000 4802

### Reason 3:
We can see that for all such events for engagement, there is a significant decrease observed except search_autocomplete. Auto_complete happens when that search has been already looked for on eLinks. For new searches, engagements have been decreased. This can be a reason for the users not preferring eLinks due to bad results or bad user interfaces but sticking to it for old searches. The same reason can be concluded with the reason 1, where the users donot clickthrough on the emails sent in the weekly digest.

##### What are the findings regarding whether drop in user activity is significant or no. What seems like the most likely cause of the drop in engagement?

There is significant difference found in  number of engagements for the last month.

The main reason for that can be reason 1 and reason 3. Also they are both correlated but different.

##### If there are questions that I can't answer using data alone, how would I go about answering them (hypothetically, assuming I actually worked at this company)? What, if anything, should the company do in response?

Such questions would be taking a review from my customers about the user interface and quality of service being offered. This will tell us how the users feel about our weekly_digest as well if they are being sent too trivial and frequently. Also, if I was working for this company this then I would be aware about the competitors and the difference in services which are being offered by them and will work on that data.

ELinks should analyze what are the important events that should be sent in the weekly_digest. Provide some discounts for newly joined users as there number is increasing but their engagament is not increasing. Improve and modernize their user interface for thew newly joined users.

##### Conclusion:
The engagement of eLinks is decreased due to less workload in the month of August for companies, bad user interface and bad search results compared to a competitive company and too frequent weekly digest sent to users which they don't find interesting.